storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java

```diff
@@ -0,0 +1,578 @@
+                // the current scope.
+                //
+                if (fileset != null) {
+                    fileset.delete();
+                }
+
+                storageListeners.onStorageEvent(eventBuilder.withException(e).build());
+                throw e;
+            }
+
+            return null;
+        };
+
+        wrap(callable);
+    }
+
+    @Override
+    public InputStream fetchLogSegment(final RemoteLogSegmentMetadata metadata,
+                                       final int startPos) throws RemoteStorageException {
+        return fetchLogSegment(metadata, startPos, Integer.MAX_VALUE);
+    }
+
+    @Override
+    public InputStream fetchLogSegment(final RemoteLogSegmentMetadata metadata,
+                                       final int startPos,
+                                       final int endPos) throws RemoteStorageException {
+        checkArgument(startPos >= 0, "Start position must be positive", startPos);
+        checkArgument(endPos >= startPos,
+                "End position cannot be less than startPosition", startPos, endPos);
+
+        return wrap(() -> {
+
+            final LocalTieredStorageEvent.Builder eventBuilder = newEventBuilder(FETCH_SEGMENT, metadata);
+            eventBuilder.withStartPosition(startPos).withEndPosition(endPos);
+
+            try {
+                final RemoteLogSegmentFileset fileset = openFileset(storageDirectory, metadata.remoteLogSegmentId());
+
+                final InputStream inputStream = newInputStream(fileset.getFile(SEGMENT).toPath(), READ);
+                inputStream.skip(startPos);
+
+                // endPosition is ignored at this stage. A wrapper around the file input stream can implement
+                // the upper bound on the stream.
+
+                storageListeners.onStorageEvent(eventBuilder.withFileset(fileset).build());
+
+                return inputStream;
+
+            } catch (final Exception e) {
+                storageListeners.onStorageEvent(eventBuilder.withException(e).build());
+                throw e;
+            }
+        });
+    }
+
+    @Override
+    public InputStream fetchIndex(RemoteLogSegmentMetadata metadata, IndexType indexType) throws RemoteStorageException {
+        EventType eventType = getEventTypeForFetch(indexType);
+        RemoteLogSegmentFileset.RemoteLogSegmentFileType fileType = getLogSegmentFileType(indexType);
+        return wrap(() -> {
+            final LocalTieredStorageEvent.Builder eventBuilder = newEventBuilder(eventType, metadata);
+
+            try {
+                final RemoteLogSegmentFileset fileset = openFileset(storageDirectory, metadata.remoteLogSegmentId());
+
+                File file = fileset.getFile(fileType);
+                final InputStream inputStream = (fileType.isOptional() && !file.exists()) ?
+                        new ByteArrayInputStream(new byte[0]) : newInputStream(file.toPath(), READ);
```

Comment:
Based on latest [RSM docs](https://github.com/apache/kafka/pull/13923) files not found, like transactional index that is optional, should return `RemoteResourceNotFound`

Reply:
Will fix this as part of https://issues.apache.org/jira/browse/KAFKA-14993 ticket